### ISSUE TITLE
Failed in start sample liblogging.virtual.so in tests/sandbox/sandbox.sh

### ISSUE THREAD

**Issue by DavidZhongSyd** [microsoft/CCF#3436]
**Describe the bug**
I followed the steps in https://microsoft.github.io/CCF/main/build_apps/demo.html, trying to run the sample liblogging.virtual.so using tests/sandbox/sandbox.sh. But it pops up many python error.

```
root@bfb4c5ad0c29:/home/CCF/build# ../tests/sandbox/sandbox.sh -p ./liblogging.virtual.so
Setting up Python environment...
Python environment successfully setup
[23:17:17.266] Starting 1 CCF node...
[23:17:17.267] Virtual mode enabled
Traceback (most recent call last):
  File "/home/CCF/tests/infra/network.py", line 1196, in network
    yield net
  File "/home/CCF/tests/sandbox/../start_network.py", line 69, in run
    network.start_and_join(args)
  File "/home/CCF/tests/infra/network.py", line 422, in start_and_join
    primary = self._start_all_nodes(args)
  File "/home/CCF/tests/infra/network.py", line 305, in _start_all_nodes
    node.start(
  File "/home/CCF/tests/infra/node.py", line 177, in start
    self._start(
  File "/home/CCF/tests/infra/node.py", line 293, in _start
    self.remote.get_startup_files(self.common_dir)
  File "/home/CCF/tests/infra/remote.py", line 855, in get_startup_files
    self.remote.get(self.pem, dst_path)
  File "/home/CCF/tests/infra/remote.py", line 450, in get
    raise ValueError(path)
ValueError: /home/CCF/build/workspace/sandbox_0/0.pem

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/CCF/tests/sandbox/../start_network.py", line 163, in <module>
    run(args)
  File "/home/CCF/tests/sandbox/../start_network.py", line 114, in run
    LOG.info("Stopping all CCF nodes...")
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/CCF/tests/infra/network.py", line 1210, in network
    net.stop_all_nodes(skip_verification=True)
  File "/home/CCF/tests/infra/network.py", line 595, in stop_all_nodes
    raise NodeShutdownError("Fatal error found during node shutdown")
infra.network.NodeShutdownError: Fatal error found during node shutdown
```

**To Reproduce**
[run in local machine]
git clone https://github.com/microsoft/CCF.git
cd CCF
create Dockerfile with content https://github.com/microsoft/CCF/blob/main/docker/ccf_ci
docker build -t ccf:1.0 .
docker run -t -i ccf:1.0 /bin/bash

[run in docker container]
cd /home/
git clone https://github.com/microsoft/CCF.git
cd CCF
cd getting_started/setup_vm/
./run.sh app-dev.yml
cd ../..
mkdir build
cd build
cmake ..
(why I install below packages, please refer to https://github.com/microsoft/CCF/issues/3426)
sudo apt-get install libc++abi1-10
sudo apt-get install libc++1-10
sudo apt-get install libuv1
sudo apt-get install libc++-10-dev
sudo apt-get install libc++abi-10-dev    
make

cd /home/
mkdir ccf_deb
cd ccf_deb
wget https://github.com/microsoft/CCF/releases/download/ccf-1.0.16/ccf_1.0.16_amd64.deb
apt install ./ccf_1.0.16_amd64.deb
```
run test
root@bc8570b00f15:/home/installationFiles# /opt/ccf/bin/cchost --version
CCF host: ccf-1.0.16
```
sudo apt-get install gcc libpq-dev -y
sudo apt-get install python-dev  python-pip -y
sudo apt-get install python3-dev python3-pip python3-venv python3-wheel -y
pip3 install wheel
sudo apt-get install curl

```
run test
root@bc8570b00f15:/home/installationFiles# /opt/ccf/bin/sandbox.sh
Setting up Python environment...
No package/app specified. Defaulting to installed JS logging app
Python environment successfully setup
[17:27:59.111] Starting 1 CCF node...
[17:27:59.111] Virtual mode enabled
[17:28:05.138] Started CCF network with the following nodes:
[17:28:05.138]   Node [0] = https://127.0.0.1:8000
[17:28:05.138] You can now issue business transactions to the /opt/ccf/bin/../lib/libjs_generic application
[17:28:05.139] Loaded JS application: /opt/ccf/bin/../samples/logging/js
[17:28:05.140] Keys and certificates have been copied to the common folder: /home/installationFiles/workspace/sandbox_common
[17:28:05.140] See https://microsoft.github.io/CCF/main/use_apps/issue_commands.html for more information
[17:28:05.141] Press Ctrl+C to shutdown the network
```
cd /home/CCF/build/
cp /home/CCF/build/samples/apps/logging/liblogging.virtual.so ./
../tests/sandbox/sandbox.sh -p ./liblogging.virtual.so

**Expected behavior**
run a simple one node test network on a single machine as what it shows in https://microsoft.github.io/CCF/main/build_apps/demo.html

```
$ cd CCF/build

$ ../tests/sandbox.sh -p ./liblogging.virtual.so
Setting up Python environment...
Python environment successfully setup
[16:14:05.294] Starting 1 CCF node...
[16:14:05.295] Virtual mode enabled
[16:14:10.010] Started CCF network with the following nodes:
[16:14:10.011]   Node [0] = https://127.0.0.1:8000
[16:14:10.011] You can now issue business transactions to the ./liblogging.virtual.so application.
[16:14:10.011] Keys and certificates have been copied to the common folder: /data/src/CCF/build/workspace/sandbox_common
[16:14:10.011] See https://microsoft.github.io/CCF/main/use_apps/issue_commands.html for more information.
[16:14:10.011] Press Ctrl+C to shutdown the network.
```

**Environment information**
Distributor ID: Ubuntu
Description: Ubuntu 20.04.3 LTS
Release: 20.04
Codename: focal

**Additional context**
May I ask if the document on https://microsoft.github.io/CCF/main/ has not been updated for a while? I feel a bit hard to follow these document.

**Comment by DavidZhongSyd** [comment]
Also failed in using /opt/ccf/bin/sandbox.sh with the similar error as tests/sandbox/sandbox.sh

```
root@bfb4c5ad0c29:/home/CCF/build# /opt/ccf/bin/sandbox.sh -p /home/CCF/build/samples/apps/logging/liblogging.virtual.so
Setting up Python environment...
Python environment successfully setup
[00:14:18.966] Starting 1 CCF node...
[00:14:18.966] Virtual mode enabled
Traceback (most recent call last):
  File "/opt/ccf/bin/infra/network.py", line 1016, in network
    yield net
  File "/opt/ccf/bin/start_network.py", line 55, in run
    network.start_and_join(args)
  File "/opt/ccf/bin/infra/network.py", line 402, in start_and_join
    primary = self._start_all_nodes(args)
  File "/opt/ccf/bin/infra/network.py", line 286, in _start_all_nodes
    node.start(
  File "/opt/ccf/bin/infra/node.py", line 128, in start
    self._start(
  File "/opt/ccf/bin/infra/node.py", line 241, in _start
    self.remote.get_startup_files(self.common_dir)
  File "/opt/ccf/bin/infra/remote.py", line 775, in get_startup_files
    self.remote.get(self.pem, dst_path)
  File "/opt/ccf/bin/infra/remote.py", line 454, in get
    raise ValueError(path)
ValueError: /home/CCF/build/workspace/sandbox_0/0.pem

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ccf/bin/start_network.py", line 144, in <module>
    run(args)
  File "/opt/ccf/bin/start_network.py", line 98, in run
    LOG.info("Stopping all CCF nodes...")
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/opt/ccf/bin/infra/network.py", line 1028, in network
    net.stop_all_nodes(skip_verification=True)
  File "/opt/ccf/bin/infra/network.py", line 580, in stop_all_nodes
    raise NodeShutdownError("Fatal error found during node shutdown")
infra.network.NodeShutdownError: Fatal error found during node shutdown
```

**Comment by DavidZhongSyd** [comment]
This error also occurs in ccfciteam/ccf-app-ci:2.0.0-dev7.

Did I miss some steps?

[Local terminal]
docker pull ccfciteam/ccf-app-ci:2.0.0-dev7
sudo docker run -it ccfciteam/ccf-app-ci:2.0.0-dev7

```
root@77fc8dab879b:~/CCF/samples/apps/logging/build# /opt/ccf/bin/sandbox.sh
Setting up Python environment...
No package/app specified. Defaulting to installed JS logging app
Python environment successfully setup
[03:19:31.848] Starting 1 CCF node...
[03:19:31.849] Virtual mode enabled
Traceback (most recent call last):
  File "/opt/ccf/bin/start_network.py", line 163, in <module>
    run(args)
  File "/opt/ccf/bin/start_network.py", line 69, in run
    network.start_and_join(args)
  File "/opt/ccf/bin/infra/network.py", line 417, in start_and_join
    primary = self._start_all_nodes(args)
  File "/opt/ccf/bin/infra/network.py", line 349, in _start_all_nodes
    primary, _ = self.find_primary(
  File "/opt/ccf/bin/infra/network.py", line 763, in find_primary
    raise PrimaryNotFound
infra.network.PrimaryNotFound

```

**Comment by eddyashton** [comment]
Hi @DavidZhongSyd. It looks like there is a problem with the `2.0.0-dev7` image - thanks for raising, we've found the problem and will resolve it for the next dev release

I was able to run the sandbox, and build and run the sample app, in the `1.0.17` released image. I'll add the instructions here.

**Comment by DavidZhongSyd** [comment]
Thanks @eddyashton, all works good except there is a typos of wrong command for "# GETing the entry we just POSTed". When it changed to the below one, it all works. 

Besides that, I noticed each time the `/opt/ccf/bin/sandbox.sh` is to be executed. the current directory need to change to `cd ~/ccf_run`. Does the sandbox.sh requires the resource under ~/ccf_run/workspace? Thanks.

```
# GETing the entry we just POSTed
curl https://127.0.0.1:8000/app/log/private?id=42 -X GET --cacert networkcert.pem --cert user0_cert.pem --key user0_privk.pem
```

**Comment by eddyashton** [comment]
@DavidZhongSyd: No, the `sandbox.sh` can be run from anywhere. It creates and fills the `workspace/` directory wherever it is run. I tend to do that in `~/ccf_run` or similar (arbitrary name, arbitrary location) to confirm it's not depending on anything unexpected in the current path.

**Comment by DavidZhongSyd** [comment]
Got it. Thanks @eddyashton
